From ee3728143afb454f41ef7f14ec0d12be596c62e3 Mon Sep 17 00:00:00 2001
From: Paul Naoumenko <paulx.naoumenko@intel.com>
Date: Fri, 11 Oct 2013 18:49:10 +0200
Subject: [PORT FROM MAIN] P2P: avoid unexpected aging of p2p devices

BZ: 107045

If expiration callback is triggered when a p2p group has been created
but no client is connected yet, the "invited" client may be considered
as lost because its last_seen timestamp is too old (whereas it sent an
invitation response few seconds ago).

This patch updates the last_seen timestamp of peer each time a p2p action
is received from it (e.g. invitation response).

Change-Id: I94185b75383c888f86b27881b064d6b26a73a47c
Orig-Change-Id: Ie80c3e4eaa3d7c4b62a3348fa9b39bce2a7930d8
Category: device enablement
Domain: CWS.WIFI-Common
Origin: upstream
Upstream-Candidate: yes
Signed-off-by: Cedric IZOARD <cedricx.izoard@intel.com>
---
 src/p2p/p2p.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/src/p2p/p2p.c b/src/p2p/p2p.c
index 738436c..a3a83bf 100644
--- a/src/p2p/p2p.c
+++ b/src/p2p/p2p.c
@@ -1652,11 +1652,17 @@ static void p2p_rx_p2p_action(struct p2p_data *p2p, const u8 *sa,
 			      const u8 *data, size_t len, int rx_freq)
 {
 	p2p_dbg(p2p, "RX P2P Public Action from " MACSTR, MAC2STR(sa));
+	struct p2p_device *dev;
+
 	wpa_hexdump(MSG_MSGDUMP, "P2P: P2P Public Action contents", data, len);
 
 	if (len < 1)
 		return;
 
+	dev = p2p_get_device(p2p, sa);
+	if (dev)
+		os_get_time(&dev->last_seen);
+
 	switch (data[0]) {
 	case P2P_GO_NEG_REQ:
 		p2p_process_go_neg_req(p2p, sa, data + 1, len - 1, rx_freq);
-- 
1.7.4.1

